lib/class-wp-sourced-attributes-block-parser.php

```diff
@@ -0,0 +1,200 @@
+<?php
+/**
+ * Block Serialization Parser
+ *
+ * @package Gutenberg
+ */
+
+/**
+ * Class WP_Sourced_Attributes_Block_Parser
+ *
+ * Parses a document and constructs a list of parsed block objects
+ *
+ * @since 6.9.0
+ */
+class WP_Sourced_Attributes_Block_Parser extends WP_Block_Parser {
+
+	/**
+	 * Parses a document and returns a list of block structures
+	 *
+	 * When encountering an invalid parse will return a best-effort
+	 * parse. In contrast to the specification parser this does not
+	 * return an error on invalid inputs.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param string                 $document Input document being parsed.
+	 * @param WP_Block_Type_Registry $registry Block type registry from which
+	 *                                         block attributes schema can be
+	 *                                         retrieved.
+	 * @param int|null               $post_id  Optional post ID.
+	 * @return WP_Block_Parser_Block[]
+	 */
+	function parse( $document, $registry = null, $post_id = null ) {
+		if ( is_null( $registry ) ) {
+			$registry = WP_Block_Type_Registry::get_instance();
+		}
+
+		$blocks = parent::parse( $document );
+
+		foreach ( $blocks as $i => $block ) {
+			$block_type = $registry->get_registered( $block['blockName'] );
+			if ( is_null( $block_type ) || ! isset( $block_type->attributes ) ) {
+				continue;
+			}
+
+			$sourced_attributes = $this->get_sourced_attributes(
+				$block,
+				$block_type->attributes,
+				$post_id
+			);
+
+			$blocks[ $i ]['attrs'] = array_merge( $block['attrs'], $sourced_attributes );
+		}
+
+		return $blocks;
+	}
+
+	/**
+	 * Returns an array of sourced attribute values for a block.
+	 *
+	 * @param WP_Block_Parser_Block $block             Parsed block object.
+	 * @param array                 $attributes_schema Attributes of registered
+	 *                                                 block type for block.
+	 * @param int|null              $post_id           Optional post ID.
+	 * @return array                                   Sourced attribute values.
+	 */
+	function get_sourced_attributes( $block, $attributes_schema, $post_id ) {
+		$attributes = array();
+
+		foreach ( $attributes_schema as $key => $attribute_schema ) {
+			if ( isset( $attribute_schema['source'] ) ) {
+				$attributes[ $key ] = $this->get_sourced_attribute(
+					$block,
+					$attribute_schema,
+					$post_id
+				);
+			}
+		}
+
+		return $attributes;
+	}
+
+	/**
+	 * Returns a sourced attribute value for a block, for attribute type which
+	 * sources from HTML.
+	 *
+	 * @param WP_Block_Parser_Block $block            Parsed block object.
+	 * @param array                 $attribute_schema Attribute schema for
+	 *                                                individual attribute to
+	 *                                                be parsed.
+	 * @return mixed                                  Sourced attribute value.
+	 */
+	function get_html_sourced_attribute( $block, $attribute_schema ) {
```

Comment:
This call is going to be expensive from a compute level. Is there anyway we can cache the result?

Reply:
> This call is going to be expensive from a compute level. Is there anyway we can cache the result?

I'd thought a bit about what might make sense to cache. Since the HTML of each block will likely be unique, I don't know that we would want to cache either the loaded HTML or the queried results, as the cache hit ratio would be very low.

What might make sense, depending on whether it makes a measurable difference:

- Caching the `$document` itself.
   - If constructing `DOMDocument` is expensive (I don't know that it is)
- Caching the converted XPath selectors
   - The conversion may or may not be expensive (might also depend which implementation we choose), but there's a higher likelihood we would reuse those on a a per-block-type basis (e.g. every paragraph will be running the `p` selector).

Reply:
How about cache the attributes, in say post meta?

Reply:
> How about cache the attributes, in say post meta?

Hm, I'd have to think about it more, but that does seem like a good idea. In fact, it might then make sense to run this sourcing logic when a post is saved, rather than at parse-time (the parse would just read the cached result).